README

seqorrmap : map sequence correlations onto structure
---------------------------------------------

GENERAL
seqorrmap maps sequence-derived residue-pair correlations onto a corresponding PDB structure.
Correlations can be created with the program PSICOV, for example.
It is mandatory to specify a file containing residue pair correlations and a PDB file
that contains the same sequence the correlations were derived for.
seqorrmap verifies that (adjusted) sequence numbers given in the correlation file exist
in the structure and that the amino acid types match. 
If residues in the correlation file cannot be mapped, the program warns or exits,
depending on the strictness settings (see below).
Two switches are provided to adjust alignment (transformation) of residues of 
the sequence in the correlation file with residues contained in the structure:
'firstSeqResidue' and 'firstStrResidue'.

The first residue number of the correlation file is '0' by definition and 
ususally that sequence starts with a residue that is contained in the structure.
In this case, only 'firstStrResidue' needs to be provided to obtain a correct alignment,
where 'firstResidue' is the residue position in the structure file that corresponds to the
first ('0') residue in the correlation file.
Otherwise, if the correlation sequence starts N-terminal to the structure,
'firstSeqResidue' needs to be provided additionally
(keeping in mind to start the residue count in the correlation file at '0').

It is crucial to check the output logs for mapping errors.

FILTERS
seqorrmap provides a range of selection filters to focus the analysis on specific
sets of correlations. 
Correlations can be filtered by the following selection criteria: 
- sequence distance between correlated residues: sLoCutoff < seq. distance < sUpCutoff
- correlation strength between correlated residues: cLoCutoff < correlation < cUpCutoff
- radius (spatial distance) between correlated residues: rLoCutoff < radius < rUpCutoff
- residue selection lists. Residue numbers in the selection lists refer to those of the structure.
  There are two residue selection lists, specified by the command line switches:
  '--select1' to select the range of first residues in correlated residue pairs, and
  '--select2' to select equivalently the range of second residues.
  An example for a residue selection looks like this:
  --select1 A:5-6 A:19 --select2 B:50
  This would select only correlations with residues A:5, A:6 and A:19 as first
  residues and B:50 as second residue in a residue-pair correlation.

Note that the numbering in selection lists refers to the PDB structure numbering.

ADVANCED MAPPING FOR REPLICAS
Sometimes protein assemblies are composed of repeated identical replicas,
which have the same chain ID names. To differentiate between the individual replicas 
in the mapping process, one can use the '--systemChain' switch. This triggers
the program to use an internal chain numbering system running invariably like 'A','B',...
for the given chains, where a new chain is detected during input whenever chain ID 
in the PDB file changes and/or a 'TER' flag is enccountered.
These internal chain names are then printed to the terminal 
output (stdout). When '--systemChain' is given, the user needs to use the 
internal chain names in the selection strings of '--select1' and '--select2'.

VMD
Those correlations that pass the filters are selected and printed out (--corrOut).
A Tcl script is generated to map the selected correlations onto the structure for 
visualisation (--tclOut).
The VMD program can be used to load the Tcl script.

CONSISTENCY
seqorrmap checks for the existence of the input data correlations in the target structure
on the basis of the specified sequence positions and amino acid types. 

'--strict_nr' triggers the program to exit if the residue numbers specified in
the correlation input data line (adjusted by --firstResidue)
cannot be mapped onto the target structure, otherwise the program's 
default behaviour is to skip the input data line and to continue.

'--strict_aa' triggers the program skip the input data line if the specified 
amino acids are not found in the target structure, otherwise the default behaviour is
to map the contact and print a warning about the amino acid mismatch.

OUTPUT
'seqorrmap.out' contains detailed information about all successfully mapped correlations.
'seqorrmap.err' contains detailed information about all UNsuccessfully mapped correlations.
'seqorrmap_corr.dat' contains the list of (cutoff-)selected correlations.
'seqorrmap.vmd' contains a TCL script to plot the structure and selected correlations in VMD.
'seqorr.mat' contains a symmetric matrix of corralation values scaled to the value range [-1 1].
'seqorrcomp.mat' is alike, but scaled to the value range [0 1] and expressed as complement 1 - [0 1].


Install / Uninstall
-------------------
Please read the general 'INSTALL' instructions.
The configuration option '--enable-debug' creates a debuggable binary.
The option '--enable-profiling' enables profiling with 'gprof'. 

For code documentation execute 'doxygen doxygen.cfg' in the 'src' directory.
Documentation files are created in 'doc/html' and 'doc/latex'.
The latex documentation is completed by executing 'make pdf' in the 
'doc/latex' directory, which creates 'refman.ps' and 'refman.pdf'.


Usage
-----

seqorrmap [--pdb ...] [--corr ...] [OPTIONS ...]
	   --pdbIn <PDB input>				(mode: mandatory, type: char  , default: void)
	   --corr <seqorr correlation input>		(mode: mandatory, type: char  , default: void)
	   --gappedSeq <gapped target sequence>		(mode: optional , type: char  , default: void)
	   --rLoCutoff <lower RMSD cutoff>		(mode: optional , type: float , default: 0.1 Angstrom)
	   --rUpCutoff <upper RMSD cutoff>		(mode: optional , type: float , default: FLT_MAX Angstrom)
	   --cLoCutoff <lower correlation cutoff>	(mode: optional , type: float , default: -FLT_MAX)
	   --cUpCutoff <upper correlation cutoff>	(mode: optional , type: float , default: FLT_MAX)
	   --sLoCutoff <lower sequence distance cutoff>	(mode: optional , type: int   , default: 0)
	   --sUpCutoff <upper sequence distance cutoff>	(mode: optional , type: int   , default: INT_MAX)
	   --firstSeqResidue <seq residue number>	(mode: optional , type: int   , default: 0)
	   --firstStrResidue <str residue number>	(mode: optional , type: int   , default: 0)
	   --strict_aa					(mode: optional , type: no_arg, default: off)
	   --strict_nr					(mode: optional , type: no_arg, default: off)
	   --silent					(mode: optional , type: no_arg, default: off)
	   --pdbOut <PDB output>			(mode: optional , type: char  , default: seqorrmap.pdb)
	   --corrOut <corr output>			(mode: optional , type: char  , default: seqorrmap_corr.dat)
	   --tclOut <TCL output>			(mode: optional , type: char  , default: seqorrmap.vmd)
	   --corrmatOut <matrix output>			(mode: optional , type: char  , default: seqorrmap.mat)
	   --corrcompmatOut <matrix output>			(mode: optional , type: char  , default: seqorrmap_comp.mat)
	   --select1 <residue selection list>		(mode: optional , type: int   , default: OFF (INT_MIN))
	   --select2 <residue selection list>		(mode: optional , type: int   , default: OFF (INT_MIN))
		--systemChain				(mode: optional , type: no_arg, default: off)
	   --cite					(mode: optional , type: no_arg, default: off)
	   --version					(mode: optional , type: no_arg, default: off)
	   --help

A typical execution line would look like this:
seqorrmap --pdbIn 0XYZ.pdb --corr seqorr_corr.dat --cLoCutoff 0.8

The same with logging of the stdout and stderr streams:
seqorrmap --pdbIn 0XYZ.pdb --corr seqorr_corr.dat --cLoCutoff 0.8 > >(tee stdout.log) 2> >(tee stderr.log >&2)


-------------------------------------------------------------------------------
Copyright (C) 2012-2013 Jens Kleinjung

Availability
------------
The program is made available under the GNU Public License for academic
scientific purposes, under the condition that proper acknowledgement
is made to the authors of the program in publications resulting from the use
of the program.

License
-------
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.



ChangeLog
- 1.4.3
Sun Oct 13 22:44:59 BST 2013
	Implemented combination of mapped correlations.

- 1.4.2
Fri Sep 27 14:05:05 BST 2013
	Allowing recursive mapping on duplicated domains to accommodate two selection lists.
	Enabled internal chain naming (A,B,...) independent of PDB chain ID.

- 1.4.1
Mon Sep 16 16:50:44 BST 2013
	Introduced two selection lists.

- 1.4.0
Tue Jun 18 16:53:28 BST 2013
	Added corrlation matrix output.
	Added more options to residue selection.

- 1.3.4
Fri May 31 11:23:59 BST 2013
	Added selection pair assertion.
	Added firstSeqResidue switch decrement.
	Corrected incorrect sequence number in output seqorrmap_corr.dat.

-1.3.3
Fri May 31 11:23:22 BST 2013
	Correction in secetion list assertion.

-1.3.2
Thu May 30 11:04:33 BST 2013
	Mismatch output goes to stderr.
	Strictness switches corrected.
	Correlations are mapped to one chain if distance difference
		to homo-oligomeric chains is marginal.

-1.3.1
Tue May 28 18:41:49 BST 2013
	Bug fix: works now with empty selection list.

-1.3.0
Tue May 28 12:16:07 BST 2013
	Printing out chain specification and atom type to VMD file.

-1.2.5
Sat May 25 09:57:31 BST 2013
	Added selection list.

-1.2.4
Tue May 21 14:33:36 BST 2013
	Improved selection filter handling.

-1.2.3
Sun May 12 20:26:54 BST 2013
	Print minimal atom distance between residues.
	Draw correlations between shortest distance in homo-oligomers.

-1.2.2
Tue Apr 30 14:55:23 BST 2013
	Added more information to correlation output file.

-1.2.1
Wed Apr 24 14:27:59 BST 2013
	VMD output respects new selection rules.

- 1.2.0
Wed Apr  3 10:51:58 BST 2013
	Allowing for gapped target sequence.

- 1.1.4
Mon Apr  1 23:43:15 BST 2013
	Selecting correlations on the basis of sequence regions.

- 1.1.3
Wed Mar 27 14:24:55 GMT 2013
	Selecting correlations on the basis of sequence distances.

- 1.1.2
Wed Mar 27 14:24:55 GMT 2013
	Improved log output, fgets behaviour, vmd output.

- 1.1.1
Tue Mar 26 10:26:02 GMT 2013
	Modified debugging output.

- 1.1.0
Thu Mar 21 15:29:20 GMT 2013
	Now reading updated correlation data format.

- 1.0.3
Wed Mar 20 11:53:51 GMT 2013
	Added switch for 'strict' behaviour.

- 1.0.2
Tue Mar 19 13:29:00 GMT 2013
	Added switch and function for TCL sctript output.

- 1.0.1 
Tue Mar 12 16:53:20 GMT 2013
	Added switch to better control cutoff values.



src/arg.c
/*==============================================================================
arg.c : parse command line arguments
Copyright (C) 2007 Jens Kleinjung
Read the COPYING file for license information.
==============================================================================*/

#include "config.h"
#include "arg.h"

/*____________________________________________________________________________*/
/** print version */
static void print_version()
{
	fprintf(stdout, "\n%s %s\n", PACKAGE, VERSION);
}

/*____________________________________________________________________________*/
/** print header */
static void print_header()
{
    fprintf(stdout, "\nseqorrmap : map sequence correlations onto structure\n");
}

/*____________________________________________________________________________*/
/** print license */
static void print_license()
{
    fprintf(stdout, "\nCopyright (C) 2012-2013 Jens Kleinjung\n"
			"You are welcome to redistribute it under certain conditions.\n"
			"Read the COPYING file for distribution details.\n\n");
}

/*____________________________________________________________________________*/
/** print citation */
static void print_citation()
{
	fprintf(stdout, "\nno citation\n");
}

/*____________________________________________________________________________*/
/** set defaults */
static void set_defaults(Arg *arg, Argpdb *argpdb)
{
    arg->pdbInFileName = "";
    arg->corrInFileName = "";
    arg->gappedseqInFileName = "";
	arg->rLoCutoff = 0.1; /* lower radius cutoff in A */
	arg->rUpCutoff = FLT_MAX; /* upper radius cutoff in A */
	arg->cLoCutoff = -FLT_MAX; /* lower correlation cutoff */
	arg->cUpCutoff = FLT_MAX; /* upper correlation cutoff */
	arg->sLoCutoff = 0; /* lower sequence distance cutoff */
	arg->sUpCutoff = INT_MAX; /* upper sequence distance cutoff */
	arg->firstSeqResidue = 0; /* correlation map starts at this sequence residue */
	arg->firstStrResidue = 0; /* correlation map starts at this structure residue */
    arg->pdbOutFileName = "seqorrmap.pdb";
    arg->corrOutFileName = "seqorrmap_corr.dat";
    arg->tclOutFileName = "seqorrmap.vmd";
    arg->corrmatOutFileName = "seqorrmap.mat";
    arg->corrcompmatOutFileName = "seqorrmap_comp.mat";
    arg->logoOutFileName = "seqorrmap.out";
    arg->logeOutFileName = "seqorrmap.err";
	arg->select = (char **)safe_malloc(2 * sizeof(char *)); /* 2 selection lists */
	arg->select[0] = ""; /* selection list */
	arg->select[1] = "";
    arg->strict_aa = 0; /* from here: run mode flags */
    arg->strict_nr = 0;
    arg->silent = 0;
    arg->multiModel = 0;
	arg->gapped = 0;
	arg->systemChain = 0; /* systematic intra-program chain numbering */
}

/*____________________________________________________________________________*/
/** check input */
static void check_input(Arg *arg, Argpdb *argpdb)
{
	if (strlen(arg->pdbInFileName) == 0)
		Error("Invalid PDB file name");
	if (strlen(arg->corrInFileName) == 0)
		Error("Invalid correlation file name");
	assert(arg->rLoCutoff >= 0. && "cutoff radius too small");
	assert(arg->rUpCutoff >= 0. && "cutoff radius too small");
	assert(arg->rUpCutoff > 0  && "sequence distance cutoff too small");
	assert(arg->rLoCutoff < arg->rUpCutoff && "lower RMSD cutoff should be smaller than upper cutoff");
	assert(arg->cLoCutoff < arg->cUpCutoff && "lower correlation cutoff should be smaller than upper cutoff");
	assert(arg->sLoCutoff < arg->sUpCutoff && "lower sequence distance cutoff should be smaller than upper cutoff");
}

/*____________________________________________________________________________*/
/** print command line arguments */
static void print_args(Arg *arg, Argpdb *argpdb)
{
    time_t now;
    time(&now);

	if (! arg->silent) {
		fprintf(stdout, "\tdate: %s", ctime(&now));

		fprintf(stdout, \
						"\nRun parameters\n"
						"\tpdbIn: %s\n"
						"\tcorr: %s\n"
						"\tgappedSeq: %s\n"
						"\trLoCutoff: %8.2e\n"
						"\trUpCutoff: %8.2e\n"
						"\tcLoCutoff: %8.2e\n"
						"\tcUpCutoff: %8.2e\n"
						"\tsLoCutoff: %d\n"
						"\tsUpCutoff: %d\n"
						"\tfirstSeqResidue: %d\n"
						"\tfirstStrResidue: %d\n"
						"\tselect[0]: %s\n"
						"\tselect[1]: %s\n"
						"\tpdbOut: %s\n"
						"\tcorrOut: %s\n"
						"\tcorrmatOut: %s\n"
						"\tcorrcompmatOut: %s\n"
						"\ttclOut: %s\n"
						"\tstrict_aa: %d\n"
						"\tstrict_nr: %d\n"
						"\tsystemChain: %d\n",
			arg->pdbInFileName, arg->corrInFileName, arg->gappedseqInFileName,
			arg->rLoCutoff, arg->rUpCutoff,
			arg->cLoCutoff, arg->cUpCutoff,
			arg->sLoCutoff, arg->sUpCutoff,
			arg->firstSeqResidue, arg->firstStrResidue,
			arg->select[0], arg->select[1],
			arg->pdbOutFileName, arg->corrOutFileName,
			arg->tclOutFileName,
			arg->corrmatOutFileName, arg->corrcompmatOutFileName,
			arg->strict_aa, arg->strict_nr,
			arg->systemChain);

		fflush(stdout);
	}
}

/*____________________________________________________________________________*/
/** parse command line long_options */
int parse_args(int argc, char **argv, Arg *arg, Argpdb *argpdb)
{
	int c;
	const char usage[] = "\nseqorrmap [--pdb ...] [--corr ...] [OPTIONS ...]\n\
	   --pdbIn <PDB input>\t\t\t\t(mode: mandatory, type: char  , default: void)\n\
	   --corr <seqorr correlation input>\t\t(mode: mandatory, type: char  , default: void)\n\
	   --gappedSeq <gapped target sequence>\t\t(mode: optional , type: char  , default: void)\n\
	   --rLoCutoff <lower RMSD cutoff>\t\t(mode: optional , type: float , default: 0.1 Angstrom)\n\
	   --rUpCutoff <upper RMSD cutoff>\t\t(mode: optional , type: float , default: FLT_MAX Angstrom)\n\
	   --cLoCutoff <lower correlation cutoff>\t(mode: optional , type: float , default: -FLT_MAX)\n\
	   --cUpCutoff <upper correlation cutoff>\t(mode: optional , type: float , default: FLT_MAX)\n\
	   --sLoCutoff <lower sequence distance cutoff>\t(mode: optional , type: int   , default: 0)\n\
	   --sUpCutoff <upper sequence distance cutoff>\t(mode: optional , type: int   , default: INT_MAX)\n\
	   --firstSeqResidue <seq residue number>\t(mode: optional , type: int   , default: 0)\n\
	   --firstStrResidue <str residue number>\t(mode: optional , type: int   , default: 0)\n\
	   --strict_aa\t\t\t\t\t(mode: optional , type: no_arg, default: off)\n\
	   --strict_nr\t\t\t\t\t(mode: optional , type: no_arg, default: off)\n\
	   --silent\t\t\t\t\t(mode: optional , type: no_arg, default: off)\n\
	   --pdbOut <PDB output>\t\t\t(mode: optional , type: char  , default: seqorrmap.pdb)\n\
	   --corrOut <corr output>\t\t\t(mode: optional , type: char  , default: seqorrmap_corr.dat)\n\
	   --tclOut <TCL output>\t\t\t(mode: optional , type: char  , default: seqorrmap.vmd)\n\
	   --corrmatOut <matrix output>\t\t\t(mode: optional , type: char  , default: seqorrmap.mat)\n\
	   --corrcompmatOut <matrix output>\t\t(mode: optional , type: char  , default: seqorrmap_comp.mat)\n\
	   --select1 <first-residue selection list>\t(mode: optional , type: int   , default: OFF (INT_MIN))\n\
	   --select2 <second-residue selection list>\t(mode: optional , type: int   , default: OFF (INT_MIN))\n\
	   --systemChain\t\t\t\t(mode: optional , type: no_arg, default: off)\n\
	   --cite\t\t\t\t\t(mode: optional , type: no_arg, default: off)\n\
	   --version\t\t\t\t\t(mode: optional , type: no_arg, default: off)\n\
	   --help\n";

    set_defaults(arg, argpdb);

    if (argc < 3) {
		print_header();
        fprintf(stderr, "%s", usage);
		print_license();
        exit(1);
    }

    /** long option definition */
    static struct option long_options[] = {
        {"pdbIn", required_argument, 0, 1},
        {"corr", required_argument, 0, 2},
        {"gappedSeq", required_argument, 0, 3},
        {"rLoCutoff", required_argument, 0, 5},
        {"rUpCutoff", required_argument, 0, 6},
        {"cLoCutoff", required_argument, 0, 7},
        {"cUpCutoff", required_argument, 0, 8},
        {"sLoCutoff", required_argument, 0, 9},
        {"sUpCutoff", required_argument, 0, 10},
        {"firstSeqResidue", required_argument, 0, 11},
        {"firstStrResidue", required_argument, 0, 12},
        {"pdbOut", required_argument, 0, 13},
        {"corrOut", required_argument, 0, 14},
        {"tclOut", required_argument, 0, 15},
        {"corrmatOut", required_argument, 0, 16},
        {"corrcompmatOut", required_argument, 0, 17},
        {"select1", required_argument, 0, 18},
        {"select2", required_argument, 0, 19},
        {"strict_aa", no_argument, 0, 20},
        {"strict_nr", no_argument, 0, 21},
        {"systemChain", no_argument, 0, 22},
        {"silent", no_argument, 0, 30},
        {"cite", no_argument, 0, 31},
        {"version", no_argument, 0, 32},
        {"help", no_argument, 0, 33},
        {0, 0, 0, 0}
    };

    /** assign parameters to long options */
    while ((c = getopt_long(argc, argv, "1:2:3:5:6:7:8:9:10:11:12:13:14:15:16:17:18:19: 20 21 22 30 31 32 33", long_options, NULL)) != -1) {
        switch(c) {
            case 1:
                arg->pdbInFileName = optarg;
                break;
            case 2:
                arg->corrInFileName = optarg;
                break;
            case 3:
                arg->gappedseqInFileName = optarg;
                break;
            case 5:
                arg->rLoCutoff = atof(optarg);
                break;
            case 6:
                arg->rUpCutoff = atof(optarg);
                break;
            case 7:
                arg->cLoCutoff = atof(optarg);
                break;
            case 8:
                arg->cUpCutoff = atof(optarg);
                break;
            case 9:
                arg->sLoCutoff = atoi(optarg);
                break;
            case 10:
                arg->sUpCutoff = atoi(optarg);
                break;
            case 11:
                arg->firstSeqResidue = atoi(optarg);
                break;
            case 12:
                arg->firstStrResidue = atoi(optarg);
                break;
            case 13:
                arg->pdbOutFileName = optarg;
                break;
            case 14:
                arg->corrOutFileName = optarg;
                break;
            case 15:
                arg->tclOutFileName = optarg;
                break;
            case 16:
                arg->corrmatOutFileName = optarg;
                break;
            case 17:
                arg->corrcompmatOutFileName = optarg;
                break;
            case 18:
                arg->select[0] = optarg;
                break;
            case 19:
                arg->select[1] = optarg;
                break;
            case 20:
                arg->strict_aa = 1;
                break;
            case 21:
                arg->strict_nr = 1;
                break;
            case 22:
                arg->systemChain = 1;
                break;
            case 30:
                arg->silent = 1;
                break;
            case 31:
                print_citation();
                exit(0);
            case 32:
				print_version();
				print_license();
                exit(0);
            case 33:
                fprintf(stderr, "%s", usage);
				print_license();
                exit(0);
            default:
                fprintf(stderr, "%s", usage);
				print_license();
                exit(1);
        }
    }

	check_input(arg, argpdb);
	if (! arg->silent) {
		print_header();
		print_args(arg, argpdb);
	}

	if (strlen(arg->gappedseqInFileName) > 0)
		++ arg->gapped;

    return 0;
}

